File: packages/skillet/src/eval/commands/serve.ts
import fs from "node:fs";
import http from "node:http";
import path from "node:path";
import pc from "picocolors";
import { findProjectRoot, resolveSkillPaths } from "../config.js";
import { writeDashboard } from "../report/html-reporter.js";
import { exitWithMissingEvalsFile } from "../utils/cli-error.js";

interface ServeOpts {
	skill: string;
	evals?: string;
	port?: string;
}

const MIME: Record<string, string> = {
	".html": "text/html",
	".json": "application/json",
	".js": "application/javascript",
	".css": "text/css",
};

export async function runServe(opts: ServeOpts) {
	const port = Number(opts.port ?? 3000);
	const paths = resolveSkillPaths(opts.skill, opts.evals);
	const skillArg = opts.skill || ".";
	const resultsRoot = path.join(findProjectRoot(paths.skillDir), ".skillet-evals", "results");
	const shouldServeAllSkills =
		!opts.evals && !fs.existsSync(paths.skillFile) && fs.existsSync(resultsRoot);
	const resultsDir = shouldServeAllSkills ? resultsRoot : paths.resultsDir;

	if (opts.evals && !fs.existsSync(paths.evalsFile)) {
		exitWithMissingEvalsFile("serve", skillArg, paths.evalsFile, true);
	}

	if (!fs.existsSync(resultsDir)) {
		fs.mkdirSync(resultsDir, { recursive: true });
	}

	writeDashboard(resultsDir);

	let skillName = shouldServeAllSkills ? "All skills" : path.basename(paths.skillDir);
	if (!shouldServeAllSkills) {
		try {
			const raw = JSON.parse(fs.readFileSync(paths.evalsFile, "utf-8"));
			if (raw.skill_name) skillName = raw.skill_name;
		} catch {
			// evals.json may not exist yet; fall back to directory name
		}
	}

	function listRunFiles(baseDir: string): string[] {
		if (shouldServeAllSkills) {
			return fs
				.readdirSync(baseDir, { withFileTypes: true })
				.filter((entry) => entry.isDirectory())
				.flatMap((entry) =>
					fs
						.readdirSync(path.join(baseDir, entry.name))
						.filter((file) => file.endsWith(".json") && file !== "latest.json")
						.map((file) => `${entry.name}/${file}`),
				)
				.sort()
				.reverse();
		}

		return fs
			.readdirSync(baseDir)
			.filter((f) => f.endsWith(".json") && f !== "latest.json")
			.sort()
			.reverse();
	}

	function resolveRequestedPath(baseDir: string, pathname: string): string | null {
		const relativePath = pathname.replace(/^\/+/, "");
		if (!relativePath) return path.join(baseDir, "index.html");
		const resolvedPath = path.resolve(baseDir, relativePath);
		if (!resolvedPath.startsWith(path.resolve(baseDir))) return null;
		return resolvedPath;
	}

	const server = http.createServer((req, res) => {
		const url = new URL(req.url ?? "/", `http://localhost:${port}`);
		const pathname = url.pathname;

		if (pathname === "/api/runs") {
			const files = listRunFiles(resultsDir);
			res.writeHead(200, {
				"Content-Type": "application/json",
				"Cache-Control": "no-cache",
			});
			res.end(JSON.stringify(files));
			return;
		}

		let filePath: string;
		if (pathname === "/" || pathname === "/index.html") {
			writeDashboard(resultsDir);
			filePath = path.join(resultsDir, "index.html");
		} else {
			const resolvedPath = resolveRequestedPath(resultsDir, pathname);
			if (!resolvedPath) {
				res.writeHead(404, { "Content-Type": "text/plain" });
				res.end("Not found");
				return;
			}
			filePath = resolvedPath;
		}

		if (fs.existsSync(filePath) && fs.statSync(filePath).isFile()) {
			const ext = path.extname(filePath);
			res.writeHead(200, {
				"Content-Type": MIME[ext] ?? "text/plain",
				"Cache-Control": ext === ".json" ? "no-cache" : "max-age=60",
			});
			fs.createReadStream(filePath).pipe(res);
			return;
		}

		res.writeHead(404, { "Content-Type": "text/plain" });
		res.end("Not found");
	});

	server.listen(port, () => {
		console.log(pc.bold("\n  skillet eval serve\n"));
		console.log(`  Skill:    ${skillName}`);
		console.log(`  Results:  ${resultsDir}`);
		console.log("");
		console.log(`  ${pc.green("→")} ${pc.bold(`http://localhost:${port}`)}`);
		console.log(`  ${pc.dim("Press Ctrl+C to stop")}\n`);
	});
}

File: packages/skillet-eval-python/src/skill_eval/commands/serve.py
from __future__ import annotations

import json
from http.server import HTTPServer, SimpleHTTPRequestHandler
from pathlib import Path

from rich.console import Console

from skill_eval.config import resolve_skill_paths
from skill_eval.report.html_reporter import write_dashboard

MIME_TYPES = {
    ".html": "text/html",
    ".json": "application/json",
    ".js": "application/javascript",
    ".css": "text/css",
}


def _make_handler(results_dir: Path):
    class Handler(SimpleHTTPRequestHandler):
        def __init__(self, *args, **kwargs):
            super().__init__(*args, directory=str(results_dir), **kwargs)

        def log_message(self, format, *args):
            pass

        def do_GET(self):
            from urllib.parse import urlparse

            parsed = urlparse(self.path)
            pathname = parsed.path

            if pathname == "/api/runs":
                files = sorted(
                    [f.name for f in results_dir.iterdir() if f.suffix == ".json" and f.name != "latest.json"],
                    reverse=True,
                )
                self.send_response(200)
                self.send_header("Content-Type", "application/json")
                self.send_header("Cache-Control", "no-cache")
                self.end_headers()
                self.wfile.write(json.dumps(files).encode())
                return

            if pathname in ("/", "/index.html"):
                write_dashboard(str(results_dir))
                pathname = "/index.html"

            file_path = results_dir / Path(pathname).name
            if file_path.exists() and file_path.is_file():
                ext = file_path.suffix
                self.send_response(200)
                self.send_header("Content-Type", MIME_TYPES.get(ext, "text/plain"))
                self.send_header(
                    "Cache-Control",
                    "no-cache" if ext == ".json" else "max-age=60",
                )
                self.end_headers()
                self.wfile.write(file_path.read_bytes())
                return

            self.send_response(404)
            self.send_header("Content-Type", "text/plain")
            self.end_headers()
            self.wfile.write(b"Not found")

    return Handler


def run_serve(skill: str, evals_path: str | None = None, port: int = 3000) -> None:
    paths = resolve_skill_paths(skill, evals_path)
    results_dir = Path(paths["results_dir"])

    if not results_dir.exists():
        results_dir.mkdir(parents=True, exist_ok=True)

    write_dashboard(str(results_dir))

    skill_name = Path(paths["skill_dir"]).name
    evals_file = Path(paths["evals_file"])
    if evals_file.exists():
        try:
            raw = json.loads(evals_file.read_text())
            if raw.get("skill_name"):
                skill_name = raw["skill_name"]
        except (json.JSONDecodeError, KeyError):
            pass

    handler = _make_handler(results_dir)
    server = HTTPServer(("", port), handler)

    console = Console()
    console.print("\n  [bold]skillet-eval serve[/bold]\n")
    console.print(f"  Skill:    {skill_name}")
    console.print(f"  Results:  {results_dir}")
    console.print()
    console.print(f"  [green]→[/green] [bold]http://localhost:{port}[/bold]")
    console.print("  [dim]Press Ctrl+C to stop[/dim]\n")

    try:
        server.serve_forever()
    except KeyboardInterrupt:
        server.shutdown()
